File: remocra/src/main/webapp/js/app/remocra/model/Oldeb.js
Ext.require('Sdis.Remocra.network.RemocraRest');
Ext.require('Sdis.Remocra.model.Commune');
Ext.require('Sdis.Remocra.model.TypeOldebAcces');
Ext.require('Sdis.Remocra.model.TypeOldebZoneUrbanisme');
Ext.require('Sdis.Remocra.model.OldebLocataire');
Ext.require('Sdis.Remocra.model.OldebPropriete');
Ext.require('Sdis.Remocra.model.OldebVisite');
Ext.require('Sdis.Remocra.model.TypeOldebCaracteristique');

Ext.define('Sdis.Remocra.model.Oldeb', {
    extend: 'Ext.data.Model',

    fields: [{
        name: 'id',
        type: 'int',
        useNull: true
    }, {
        name: 'geometrie',
        type: 'string'
    }, {
        name: 'section',
        type: 'string'
    }, {
        name: 'parcelle',
        type: 'string'
    }, {
        name: 'voie',
        type: 'string'
    }, {
        name: 'volume',
        type: 'integer'
    }, {
        name: 'largeurAcces',
        type: 'integer'
    }, {
        name: 'portailElectrique',
        type: 'boolean'
    }, {
        name: 'codePortail',
        type: 'string'
    }, {
        name: 'jsonGeometrie',
        type: 'auto',
        persist: false,
        useNull: true
    }, {
        name: 'actif',
        type: 'boolean',
        defaultValue: true

    }, {
        name: 'numVoie',
        type: 'string'
    }, {
        name: 'nomZoneUrbanisme',
        type: 'string'
    }, {
        name: 'adresse',
        type: 'string'
    }, {
        name: 'avis',
        type: 'string'
    }, {
        name: 'dateDerniereVisite',
        type: 'date',
        dateFormat: 'c'
    }, {
        name: 'debroussaillement',
        type: 'string'
    }, {
        name: 'lieuDit',
        type: 'string'
    } /*
         * , { name: 'commune', type: 'fk', useNull: true, defaultValue: null }
         */],

    associations: [{
        type: 'belongsTo',
        model: 'Sdis.Remocra.model.Commune',
        associationKey: 'commune',
        getterName: 'getCommune',
        setterName: 'setCommune',
        associatedName: 'Commune',
        persist: true
    }, {
        type: 'belongsTo',
        model: 'Sdis.Remocra.model.TypeOldebAcces',
        associationKey: 'acces',
        getterName: 'getAcces',
        setterName: 'setAcces',
        associatedName: 'Acces',
        persist: true
    }, {
        type: 'belongsTo',
        model: 'Sdis.Remocra.model.TypeOldebZoneUrbanisme',
        associationKey: 'zoneUrbanisme',
        getterName: 'getZoneUrbanisme',
        setterName: 'setZoneUrbanisme',
        associatedName: 'ZoneUrbanisme',
        persist: true
    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.OldebLocataire',
        associationKey: 'oldebLocataires',
        name: 'oldebLocataires',
        persist: true,
        getterName: 'getOldebLocataires',
        setterName: 'setOldebLocataires'

    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.OldebPropriete',
        associationKey: 'oldebProprietes',
        name: 'oldebProprietes',
        getterName: 'getOldebProprietes',
        setterName: 'setOldebProprietes',
        associatedName: 'oldebProprietes',
        persist: true
    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.TypeOldebCaracteristique',
        associationKey: 'typeOldebCaracteristiques',
        name: 'typeOldebCaracteristiques',
        getterName: 'getTypeOldebCaracteristiques',
        setterName: 'setTypeOldebCaracteristiques',
        associatedName: 'typeOldebCaracteristiques',
        persist: true
    }, {
        type: 'hasMany',
        model: 'Sdis.Remocra.model.OldebVisite',
        name: 'oldebVisites',
        associationKey: 'oldebVisites',
        getterName: 'getOldebVisites',
        setterName: 'setOldebVisites',
        associatedName: 'oldebVisites',
        persist: true
    } ],

    proxy: {
        type: 'remocra.rest',
        url: Sdis.Remocra.util.Util.withBaseUrl('../oldeb')
    }
});

File: remocra/src/main/java/fr/sdis83/remocra/domain/remocra/Oldeb.java
package fr.sdis83.remocra.domain.remocra;

import com.vividsolutions.jts.geom.Geometry;
import fr.sdis83.remocra.domain.utils.RemocraDateHourTransformer;
import fr.sdis83.remocra.util.Featurable;
import fr.sdis83.remocra.util.Feature;
import java.util.Date;
import java.util.Set;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OrderBy;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.validation.constraints.NotNull;
import org.hibernate.annotations.Formula;
import org.hibernate.annotations.Type;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.roo.addon.javabean.RooJavaBean;
import org.springframework.roo.addon.jpa.activerecord.RooJpaActiveRecord;
import org.springframework.roo.addon.json.RooJson;
import org.springframework.roo.addon.tostring.RooToString;

@RooJavaBean
@RooToString
@RooJpaActiveRecord(
    versionField = "",
    finders = {"findOldebsByCommune"})
@RooJson
public class Oldeb implements Featurable {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "id")
  private Long id;

  @ManyToMany
  @JoinTable(
      name = "oldeb_caracteristique",
      joinColumns = {@JoinColumn(name = "oldeb", nullable = false)},
      inverseJoinColumns = {@JoinColumn(name = "caracteristique", nullable = false)})
  private Set<TypeOldebCaracteristique> typeOldebCaracteristiques;

  @OneToMany(mappedBy = "oldeb")
  private Set<OldebLocataire> oldebLocataires;

  @OneToMany(mappedBy = "oldeb")
  private Set<OldebPropriete> oldebProprietes;

  @OneToMany(mappedBy = "oldeb")
  @OrderBy("dateVisite DESC")
  private Set<OldebVisite> oldebVisites;

  @ManyToOne
  @JoinColumn(name = "commune", referencedColumnName = "id", nullable = false)
  private Commune commune;

  @ManyToOne
  @JoinColumn(name = "acces", referencedColumnName = "id")
  private TypeOldebAcces acces;

  @ManyToOne
  @JoinColumn(name = "zone_urbanisme", referencedColumnName = "id")
  private TypeOldebZoneUrbanisme zoneUrbanisme;

  @NotNull
  @Type(type = "org.hibernate.spatial.GeometryType")
  private Geometry geometrie;

  @Formula("ST_AsGeoJSON(geometrie)")
  private String jsonGeometrie;

  @Column(name = "section", unique = true)
  @NotNull
  private String section;

  @Column(name = "parcelle", unique = true)
  @NotNull
  private String parcelle;

  @Column(name = "num_voie")
  private String numVoie;

  @Column(name = "voie")
  private String voie;

  @Column(name = "lieu_dit")
  private String lieuDit;

  @Column(name = "volume")
  @NotNull
  private Integer volume;

  @Column(name = "largeur_acces")
  private Integer largeurAcces;

  @Column(name = "portail_electrique")
  @NotNull
  private boolean portailElectrique;

  @Column(name = "code_portail")
  private String codePortail;

  @Column(name = "actif")
  private Boolean actif;

  // Cas quand :
  @Formula(
      value =
          " CASE BTRIM(voie)"
              // voie est null
              + " WHEN null THEN null"
              // voie contient une chaine vide
              + " WHEN '' THEN null"
              // retour de la concaténation des valeurs
              + " ELSE BTRIM(COALESCE(num_voie, '') || ' ' || voie || ' ' || COALESCE(lieu_dit, '')) "
              + "END")
  private String adresse;

  // ov.date_visite desc limit 1 utilisé pour avoir la date dernière visite
  @Formula(
      "(select ov.date_visite from remocra.oldeb_visite ov where ov.oldeb = id order by ov.date_visite desc limit 1)")
  @Temporal(TemporalType.TIMESTAMP)
  @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
  private Date dateDerniereVisite;

  @Formula("(select zu.nom from remocra.type_oldeb_zone_urbanisme zu where zu.id = zone_urbanisme)")
  private String nomZoneUrbanisme;

  @Formula(
      " (select tod.nom from remocra.oldeb_visite ov inner join remocra.type_oldeb_debroussaillement tod on (ov.debroussaillement_parcelle = tod.id)"
          + "where ov.oldeb = id order by ov.date_visite desc limit 1) ")
  private String debroussaillement;

  @Formula(
      "(select toc.nom from remocra.oldeb_visite ov inner join remocra.type_oldeb_avis toc on (ov.avis = toc.id) "
          + "where ov.oldeb = id order by ov.date_visite desc limit 1)")
  private String avis;

  @Override
  public Feature toFeature() {
    Feature feature = new Feature(this.id, this.getJsonGeometrie());
    feature.addProperty("section", this.getSection());
    feature.addProperty("parcelle", this.getParcelle());
    return feature;
  }
}
